### DISCUSSION TITLE
One Certificate one client using EMQX plugin in version>=5.

### DISCUSSION THREAD

**Discussion by sc-sityad** [emqx/emqx#10776] [Q&A] [answered]
I am able to run the plugin_template that is there. But I am not able to store the client certificate in a list or something if its unique, and if its dublicate then I am not allowing the client to connect to emqx. I need client certificate of that client and compare it with existiung list. If its there or not accordingly I handled the code. 

Q1) How to store the certificate in a list or something in erlang language(Plugin).
Q2) How to retrieve certificate of the client that tries to connect.

Its a modification of on_client _connect.

Here is my code:

`-module(one_client_one_certificate_plugin).

-behaviour(emqx_hook).

-export([start/2, stop/1]).
-export([on_client_connect/3]).

start(PluginName, _Opts) ->
    emqx_plugin:reg(PluginName, ?MODULE, []),
    used_certificates_server:start_link(), % Start the used_certificates_server process
    ok.

stop(PluginName) ->
    emqx_plugin:unreg(PluginName),
    ok.

on_client_connect(ClientId, _Username, Session) ->
    case validate_client_certificate(ClientId, Session) of
        ok ->
            {ok, Session};
        {error, Reason} ->
            {error, Reason}
    end.

validate_client_certificate(ClientId, Session) ->
    case get_certificate(ClientId) of
        {ok, _Certificate} ->
            {error, <<"Duplicate client certificate detected.">>};
        {error, _} ->
            store_certificate(ClientId, Session),
            ok
    end.

get_certificate(ClientId) ->
    used_certificates_server:get_certificate(ClientId).

store_certificate(ClientId, Session) ->
    Certificate = get_client_certificate(Session), % Retrieve the client certificate
    used_certificates_server:store_certificate(ClientId, Certificate).

%% Used Certificates Server Module
-module(used_certificates_server).

-behaviour(gen_server).

-export([start_link/0, store_certificate/2, get_certificate/1]).
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).

-record(state, {used_certificates = dict:new()}).

start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

store_certificate(ClientId, Certificate) ->
    gen_server:call(?MODULE, {store_certificate, ClientId, Certificate}).

get_certificate(ClientId) ->
    gen_server:call(?MODULE, {get_certificate, ClientId}).

init([]) ->
    {ok, #state{}}.

handle_call({store_certificate, ClientId, Certificate}, _From, State) ->
    UpdatedState = State#state{used_certificates = dict:store(ClientId, Certificate, State#state.used_certificates)},
    {reply, ok, UpdatedState};

handle_call({get_certificate, ClientId}, _From, State) ->
    Certificate = dict:find(ClientId, State#state.used_certificates),
    {reply, Certificate, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%% Helper Function to Get Client Certificate from Session
get_client_certificate(Session) ->
    Certificates = session:get(certificates, Session),
    case lists:nth(1, Certificates) of
        {_, Certificate, _} ->
            Certificate;
        _ ->
            undefined
    end.
`

**Comment by lafirest** [comment, marked as answer]
See this #10777  to view the answer.
